library/common/main_interface.h

```diff
@@ -1,12 +1,141 @@
 #pragma once
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
 
 // NOLINT(namespace-envoy)
 
+/**
+ * Handle to an Envoy engine instance. Valid only for the lifetime of the engine and not intended
+ * for any external interpretation or use.
+ */
+typedef int envoy_engine_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP request. Valid only for the duration of the request and not
+ * intended for any external interpretation or use.
+ */
+typedef int envoy_request_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP stream. Valid only for the duration of the stream and not
+ * intended for any external interpretation or use.
+ */
+typedef int envoy_stream_t;
+
+/**
+ * Result codes returned by all calls made to this interface.
+ */
+typedef enum { ENVOY_SUCCESS, ENVOY_FAILURE } envoy_status_t;
+
+/**
+ * Error code associated with terminal status of a HTTP stream or request.
+ */
+typedef enum {} envoy_error_code_t;
+
+typedef struct {
+  envoy_status_t status;
+  envoy_request_t request;
+} envoy_request_pair;
+
+typedef struct {
+  envoy_status_t status;
+  envoy_stream_t stream;
+} envoy_stream_pair;
+
+typedef struct {
+  char* name;
+  char* value;
+} envoy_header;
+
+typedef struct {
+  size_t length;
+  envoy_header* headers;
+} envoy_headers;
+
+typedef struct {
+  size_t length;
+  uint8_t* bytes;
+} envoy_data;
+
+const envoy_data envoy_nodata = {0, NULL};
+
+typedef struct {
+  envoy_error_code_t error_code;
+  size_t length;
+  char* string;
+} envoy_error;
+
+#ifdef __cplusplus
+extern "C" { // function pointers
+#endif
+
+typedef void (*on_headers)(envoy_headers headers, bool end_stream);
```

Comment:
Looks like these APIs are designed for streaming. How does this translate to unary requests?

Reply:
We left this direction open so that the implementation of the observer can handle as desired. A simple implementation on top of these callbacks for a unary http request could follow a similar interface to:

https://github.com/envoyproxy/envoy/blob/master/include/envoy/http/async_client.h#L30-L44